File: python/rrt.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-

from mapinfo import MapInfo
from random import randint
from scipy.spatial import cKDTree
import math

class RRT(object):
    def __init__(self, q_init):
        self._root = q_init
        self._rrt = {q_init:q_init}

    def search_nearest_vertex(self, p):
        rrt_list = self._rrt.items()
        distance = [(q[0][0] - p[0]) ** 2 + (q[0][1] - p[1]) ** 2 for q in rrt_list]
        idx = distance.index(min(distance))
        return rrt_list[idx][0]

    def is_contain(self, q):
        return q in self._rrt

    def is_root(self, q):
        return q == self._root

    def add(self, q_new, q_near):
        self._rrt[q_new] = q_near

    def get_rrt(self):
        return self._rrt

    def get_parent(self, q):
        return self._rrt[q]

    def extend(self, q_rand, okdtree):
        # find nearest point in rrt
        q_near = self.search_nearest_vertex(q_rand)
        # calc new vertex
        q_new = self._calc_new_point(q_near, q_rand, delta_q=2.0)
        if is_collision(q_new, okdtree):
            return None
        self.add(q_new, q_near)
        return q_new

    def _calc_new_point(self, q_near, q_rand, delta_q=1.0):
        if distance(q_near, q_rand) < delta_q:
            return q_rand
        angle = math.atan2(q_rand[1] - q_near[1], q_rand[0] - q_near[0])
        q_new = (q_near[0] + math.cos(angle), q_near[1] + math.sin(angle))
        return q_new

def is_collision(p, okdtree):
    d, _ = okdtree.query(p)
    return d <= 1.0

def distance(p1, p2):
    return math.sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2)

def reconstruct_path(rrt, end):
    path = []
    q = end
    while not rrt.is_root(q):
        path.append(q)
        q = rrt.get_parent(q)
    path.append(q)
    return path

def rrt_planning(map_info, display=False):
    rrt = RRT(map_info.start)
    okdtree = cKDTree(map_info.obstacle)
    while True:
        # generate random point
        if randint(0, 10) > 2:
            q_rand = (randint(1, map_info.width - 1), randint(1, map_info.height - 1))
            if q_rand == map_info.start or q_rand in map_info.obstacle or rrt.is_contain(q_rand):
                continue
        else:
            q_rand = map_info.end
        q_new = rrt.extend(q_rand, okdtree)
        if not q_new:
            continue
        if display:
            map_info.set_rand(q_rand)
            map_info.set_rrt(rrt.get_rrt())
        # check goal
        if distance(q_new, map_info.end) <= 1.0:
            if q_new != map_info.end:
                rrt.add(map_info.end, q_new)
            return reconstruct_path(rrt, map_info.end)

if __name__ == "__main__":
    m = MapInfo(60, 40)
    m.show()
    m.start = (10, 10)
    m.end = (50, 30)
    m.obstacle = [(20, i) for i in range(30)] + [(40, 40 - i) for i in range(30)]
    raw_input('enter to start ...')
    m.path = rrt_planning(m, display=True)
    m.wait_close()

File: cpp_rviz/src/rrt.cpp
#include "rrt.h"

void RRT::set_root(KDPoint &p)
{
    _root.assign(p.begin(), p.end());
    _rrt.push_back(std::make_pair(_root, 0));
}

KDPoint RRT::SearchNearestVertex(KDPoint &q_rand)
{
    std::vector<double> d;
    for (auto pair : _rrt)
    {
        d.push_back(Distance(pair.first, q_rand));
    }
    int i = std::min_element(d.begin(), d.end()) - d.begin();
    return _rrt[i].first;
}

KDPoint RRT::CalcNewPoint(KDPoint &q_near, KDPoint &q_rand)
{
    if (Distance(q_near, q_rand) < 1.0)
    {
        return q_rand;
    }
    double angle = std::atan2(q_rand[1] - q_near[1], q_rand[0] - q_near[0]);
    double x_new = q_near[0] + std::cos(angle);
    double y_new = q_near[1] + std::sin(angle);
    KDPoint p = {x_new, y_new};
    return p;
}

void RRT::Add(KDPoint &q_new, KDPoint &q_near)
{
    int i = std::find_if(
        _rrt.begin(), _rrt.end(),
        [&](std::pair<KDPoint, int> &pair)
        {
            return (pair.first == q_near);
        }
    ) - _rrt.begin();
    _rrt.push_back(std::make_pair(q_new, i));
}

KDPoint RRT::GetParent(KDPoint &p)
{
    auto it = std::find_if(
        _rrt.begin(), _rrt.end(),
        [&](std::pair<KDPoint, int> &pair)
        {
            return (pair.first == p);
        }
    );
    return _rrt[it->second].first;
}

double RRT::Cost(KDPoint &point)
{
    double c = 0.0;
    KDPoint p = point;
    while (p != _root)
    {
        KDPoint f = GetParent(p);
        c += Distance(p, f);
        p = f;
    }
    return c;
}

void RRT::Rewire(KDPoint &p, double r, std::function<bool (KDPoint &p1, KDPoint &p2)> Collision)
{
    std::vector<KDPoint> nears;
    auto it_p = std::find_if(
        _rrt.begin(), _rrt.end(),
        [&](std::pair<KDPoint, int> &pair)
        {
            return (pair.first == p);
        }
    );
    std::for_each(
        _rrt.begin(), _rrt.end(),
        [&](std::pair<KDPoint, int> &pair)
        {
            if ((pair.first != p) && (Distance(pair.first, p) < r) && (!Collision(pair.first, p)))
            {
                nears.push_back(pair.first);
            }
        }
    );
    for (auto pt : nears)
    {
        if (Cost(pt) + Distance(pt, p) < Cost(p))
        {
            int idx = std::find_if(
                _rrt.begin(), _rrt.end(),
                [&](std::pair<KDPoint, int> &pair)
                {
                    return (pair.first == pt);
                }
            ) - _rrt.begin();
            it_p->second = idx;
        }
    }
    for (auto pt : nears)
    {
        if (Cost(p) + Distance(pt, p) < Cost(pt))
        {
            auto it_pt = std::find_if(
                _rrt.begin(), _rrt.end(),
                [&](std::pair<KDPoint, int> &pair)
                {
                    return (pair.first == pt);
                }
            );
            it_pt->second = int(it_p - _rrt.begin());
        }
    }
}
